### ISSUE TITLE
Dripsy v1 release notes

### ISSUE THREAD

**Issue by nandorojo** [nandorojo/dripsy#34]
Dripsy v1 has a number of exciting additions, some minor breaking changes, and styling differences. I'm calling it v1 so that it's clear that there are major changes to the expected web behavior – namely that it works with SSR and has an escape hatch.

To use it before it's released, use `yarn add dripsy@fresnel-2`. Please test it out and report any issues you find which aren't specified in this doc.

Dripsy for iOS and Android is basically untouched by this update. If you use Dripsy on web, this update is very relevant.

I recommend reading #19 and #14 for context.

# [feature] SSR Support

This section describes how the new web behavior works on Dripsy, and why it works for server side rendering. At the very least, check out "changes in web behavior" below.

Dripsy now fully supports SSR. Please refer to the readme for the simple installation steps.

- [x] Next.js tested
- [x] Normal RNW tested
- [ ] Gatsby tested (coming soon)

Ever since React Native Web [deprecated the classname](https://github.com/necolas/react-native-web/issues/1318), the path to responsive styles built in React Native has been a challenge. Without the `className` prop, we can only use JS dimensions to determine which media breakpoint to display. This is the right direction for the ecosystem in general – JS-based styles that are consistent on all platforms. 

However, JS-only dimensions don't work with SSR, since the server doesn't know the screen size with certainty.

The solution used in this update is to wrap every responsive element rendered on the web with a `div`. To see how this is achieved, please refer to the [docs for Fresnel](https://github.com/artsy/fresnel#why-not-conditionally-render).

This `div`'s children conditionally render depending on the how big the screen size is. The first paint, which can occur on the server, renders every breakpoint. It then uses a `className` on the `div` and hides the ones from the wrong breakpoint using CSS. Once the component has mounted, the `div` unmounts its children if their styles don't fit the current breakpoint. This is achieved thanks to @artsy's fresnel library.

# [important] Changes in web behavior

**This section is very important, if you're upgrading I recommend reading it.** 

This section discusses the `webContainerSx` prop addition. This prop is an escape hatch for the occasional unintended web behavior for flexbox and percentage widths.

**TLDR:** If you find that your `flex` or `width` values aren't working properly on web, try adding the styles to the `webContainerSx` prop.

Ok, onto the explanation.

It is important to understand what it means that we will be using `@artsy/fresnel`, as mentioned in the SSR explanation above. When we have responsive styles on web, we are wrapping our React Native element with a `div`. This means that sometimes, your styles might not show work. 

Basically, this only matters for things like percentage widths and flexbox stretching. The times it matters are rare, but they're non-zero, so I'm going to explain in detail.

Take this example. What do you think this should look like?

```jsx
<View sx={{ flexDirection: 'row', height: 500 }}>
  <View sx={{ bg: 'blue', height: '100%', flex: [1] }} />
  <View sx={{ bg: 'red', height: '100%', flex: [1] }} />
</View>
```

You might expect that it would look like this, right?

![Frame 1 (6)](https://user-images.githubusercontent.com/13172299/94039956-5c14a500-fd96-11ea-98bf-6db4319eff93.png)

Well, turns out it doesn't, exactly. 

This short video explains why: https://www.loom.com/share/997c9af614ea4b5392831bfc6db54cd0

Basically, any time you use an array style (in this case, `width: [50%]`,) Dripsy wraps your item in a `div`. While wrapping your item in a `div` doesn't usually matter, in the case of percentage widths, `flex: 1`, etc., it does. Since our blue and red boxes are both wrapped with `div`s, and the `div`s don't know to stretch the container, they don't. Thus, our `View` is `50%` width and `100%` height of a `div` whose size is unspecified.

In those cases, you can style the container div with the `webContainerSx` prop, as seen in the video.  

This is the solution for the example above:

```jsx
<View sx={{ flexDirection: 'row', height: 500 }}>
  <View webContainerSx={{ flex: [1] }} sx={{ bg: 'blue', height: '100%', flex: [1] }} />
  <View webContainerSx={{ flex: [1] }} sx={{ bg: 'red', height: '100%', flex: [1] }} />
</View>
```

Here we styled the container div for web. All set!

If you never use responsive array styles, then this won't affect you. Once a component has a single array style, though, this does start to affect how it is rendered.

Rely on `webContainerSx` if you ever find a responsive width/responsive height/flexbox value doesn't work properly on web. And that is all!

# [feature] Add `styled` function

This is a nicer API than `createThemedComponent` in my mind, and with a cooler name. They do the exact same thing though, so no need to change if you're already using `createThemedComponent`.

```jsx
import { View } from 'react-native'
import { styled } from 'dripsy' 

const StyledView = styled(View)({
  flex: 1,
  bg: 'primary'
})
```

Recursively styling Dripsy components doesn't work yet, see #33 for that. For now, you can only pass native components to `styled`.

# [feature] Add `themeKey` to components

```jsx
const theme = {
  cards: {
   primary: {...}
  }
}

...

<View 
  themeKey="cards"
  variant="primary"
/>
```

Before, we had to use the `createThemedComponent` to give a component a `themeKey`.

# [minor] `ThemeProvider` renamed to `DripsyProvider`

Change this in your imports after updating.

# [breaking] changes

- See the web behavior above.
- There will soon be breaking changes for how Dripsy is initialized at the root. This will be necessary to allow custom breakpoints. Keep an eye out for this at https://github.com/nandorojo/dripsy/issues/25

# [coming soon]

- Custom breakpoints (this will require a breaking change)
- JSX pragma that allows any component to be styled
- Custom style props for custom components that accept more than just the `style` prop
- Recursively create dripsy components

I'm also looking into custom theme key typescript suggestions still.

**Comment by cmaycumber** [comment]
Does the styled function have the ability to add the themeKey and defaultVariant in a similar fashion as the createThemeComponent API or is it intended specifically for styling purposes?

**Comment by nandorojo** [comment]
It does, it works functionally the exact same:

```js
styled(Component, { themeKey, defaultVariant })({
  ...
})
```

The second argument of `styled` is optional: https://github.com/nandorojo/dripsy/blob/master/src/css/styled.tsx

**Comment by nandorojo** [comment]
Tying to #36 to track `webContainerSx` bug.
